sdk/eventhub/azure-eventhubs/azure/eventhub/_client_base.py

```diff
@@ -0,0 +1,371 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from __future__ import unicode_literals
+
+import logging
+import uuid
+import time
+import functools
+import collections
+from typing import Any, TYPE_CHECKING
+try:
+    from urlparse import urlparse  # type: ignore
+    from urllib import urlencode, quote_plus  # type: ignore
+except ImportError:
+    from urllib.parse import urlparse, urlencode, quote_plus
+
+from uamqp import (
+    AMQPClient,
+    Message,
+    authentication,
+    constants,
+    errors,
+    compat
+)
+
+from .exceptions import _handle_exception, ClientClosedError
+from ._configuration import Configuration
+from ._utils import parse_sas_token, utc_timestamp
+from ._common import EventHubSharedKeyCredential, EventHubSASTokenCredential
+from ._connection_manager import get_connection_manager
+from ._constants import (
+    CONTAINER_PREFIX,
+    JWT_TOKEN_SCOPE,
+    MGMT_OPERATION,
+    MGMT_PARTITION_OPERATION
+)
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential  # type: ignore
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _parse_conn_str(conn_str):
+    endpoint = None
+    shared_access_key_name = None
+    shared_access_key = None
+    entity_path = None
+    for element in conn_str.split(';'):
+        key, _, value = element.partition('=')
+        if key.lower() == 'endpoint':
+            endpoint = value.rstrip('/')
+        elif key.lower() == 'hostname':
+            endpoint = value.rstrip('/')
+        elif key.lower() == 'sharedaccesskeyname':
+            shared_access_key_name = value
+        elif key.lower() == 'sharedaccesskey':
+            shared_access_key = value
+        elif key.lower() == 'entitypath':
+            entity_path = value
+    if not all([endpoint, shared_access_key_name, shared_access_key]):
+        raise ValueError(
+            "Invalid connection string. Should be in the format: "
+            "Endpoint=sb://<FQDN>/;SharedAccessKeyName=<KeyName>;SharedAccessKey=<KeyValue>")
+    return endpoint, shared_access_key_name, shared_access_key, entity_path
+
+
+def _generate_sas_token(uri, policy, key, expiry=None):
+    """Create a shared access signiture token as a string literal.
+    :returns: SAS token as string literal.
+    :rtype: str
+    """
+    from base64 import b64encode, b64decode
```

Comment:
Is there a reason we're putting these here (e.g. to catch an import error at the function level) as opposed to toplevel?

Reply:
hmm nope - I will move